packages/astro/src/core/build/static-build.ts

```diff
@@ -231,85 +235,92 @@ async function ssrBuild(
-	opts.logger.info('SKIP_FORMAT', `\n${bgGreen(black(' building client (vite) '))}`);
-
-	const viteBuildConfig: vite.InlineConfig = {
-		...viteConfig,
-		build: {
-			target: 'esnext',
-			...viteConfig.build,
-			emptyOutDir: false,
-			outDir: fileURLToPath(out),
-			copyPublicDir: ssr,
-			rollupOptions: {
-				...viteConfig.build?.rollupOptions,
-				input: Array.from(input),
-				output: {
-					format: 'esm',
-					entryFileNames: `${settings.config.build.assets}/[name].[hash].js`,
-					chunkFileNames: `${settings.config.build.assets}/[name].[hash].js`,
-					assetFileNames: `${settings.config.build.assets}/[name].[hash][extname]`,
-					...viteConfig.build?.rollupOptions?.output,
-				},
-				preserveEntrySignatures: 'exports-only',
-			},
-		},
-		plugins: [...vitePlugins, ...(viteConfig.plugins || []), ...lastVitePlugins],
-		envPrefix: viteConfig.envPrefix ?? 'PUBLIC_',
-		base: settings.config.base,
+	>();
+
+	const mutate: MutateChunk = (chunk, targets, newCode) => {
+		chunk.code = newCode;
+		mutations.set(chunk.fileName, {
+			targets,
+			code: newCode,
+		});
 	};
 
-	const updatedViteBuildConfig = await runHookBuildSetup({
-		config: settings.config,
-		pages: internals.pagesByKeys,
-		vite: viteBuildConfig,
-		target: 'client',
-		logger: opts.logger,
+	await manifestBuildPostHook(opts, internals, {
+		ssrOutputs,
+		clientOutputs,
+		prerenderOutputs,
+		mutate,
 	});
 
-	const buildResult = await vite.build(updatedViteBuildConfig);
-	return buildResult;
+	await writeMutatedChunks(opts, mutations, prerenderOutputs);
 }
 
-async function runPostBuildHooks(
-	container: AstroBuildPluginContainer,
-	ssrOutputs: vite.Rollup.RollupOutput[],
-	clientOutputs: vite.Rollup.RollupOutput[],
+async function writeMutatedChunks(
+	opts: StaticBuildOptions,
+	mutations: Map<
+		string,
+		{
+			targets: string[];
+			code: string;
+		}
+	>,
+	prerenderOutputs: vite.Rollup.RollupOutput[],
 ) {
-	const mutations = await container.runPostHook(ssrOutputs, clientOutputs);
-	const config = container.options.settings.config;
-	const build = container.options.settings.config.build;
+	const { settings } = opts;
+	const config = settings.config;
+	const build = settings.config.build;
+	const serverOutputDir = getServerOutputDirectory(settings);
+
 	for (const [fileName, mutation] of mutations) {
-		const root =
-			container.options.settings.buildOutput === 'server'
-				? mutation.targets.includes('server')
-					? build.server
-					: build.client
-				: getOutDirWithinCwd(config.outDir);
+		let root: URL;
+
+		// Check if this is a prerender file by looking for it in prerender outputs
+		const isPrerender = prerenderOutputs.some(output =>
+			output.output.some(chunk =>
+				chunk.type !== 'asset' && (chunk as any).fileName === fileName
+			)
+		);
+
+		if (isPrerender) {
+			// Write to prerender directory
+			root = new URL('./.prerender/', serverOutputDir);
```

Comment:
It would be nice to have a function that creates this directory, since it's the second `./.prenreder/` I see